src/common_modules/vsts-cli-package-common/vsts/cli/package/common/artifacttool_updater.py

```diff
@@ -0,0 +1,137 @@
+            release_id = "custom_%s" % str(uuid.uuid4()) # ensures that the custom URL is always downloaded fresh
+        else:
+            override_version = os.environ.get(self.ARTIFACTTOOL_OVERRIDE_VERSION_ENVKEY)
+            try:
+                release = self._get_current_release(team_instance, override_version)
+            except Exception as ex:
+                raise CLIError('Failed to find UPack tooling: %s' % ex) 
+            release_uri, release_id = release
+
+        # Determine the path for the release, and skip downloading if it already exists
+        logger.debug("Checking if we already have ArtifactTool release '%s'", release_id)
+        release_dir = self._get_release_dir(release_id)
+        if os.path.exists(release_dir):
+            logger.debug("Not updating ArtifactTool because the current release already exists at '%s'" % release_dir)
+            return release_dir
+              
+        # Doesn't already exist. Download and extract the release.
+        logger.debug("Updating to ArtifactTool release %s since it doesn't exist at %s" % (release_id, release_dir))
+        self._update_artifacttool(release_uri, release_dir, release_id)
+
+        return release_dir
+
+    def _get_current_release(self, team_instance, override_version):
+        connection = get_vss_connection(team_instance)
+        client = connection.get_client('vsts.cli.package.common.client_tool.client_tool_client.ClientToolClient')
+        logger.debug("Looking up current version of ArtifactTool...")
+        release = client.get_clienttool_release("ArtifactTool", os_name=platform.system(), arch=platform.machine(), version=override_version)
+        return (release.uri, self._compute_id(release)) if release is not None else None
+
+    def _update_artifacttool(self, uri, release_dir, release_id):
+        with humanfriendly.Spinner(label="Downloading UPack tooling (%s)" % release_id, total=100, stream=sys.stderr) as spinner:
+            spinner.step()
+            logger.debug("Downloading ArtifactTool from %s" % uri)
+
+            # Make the request, determine the total size
+            response = requests.get(uri, stream=True)
+            content_length_header = response.headers['Content-Length'].strip()
+            content_length = int(content_length_header)
+
+            # Do the download, updating the progress bar
+            content=io.BytesIO()
+            bytes_so_far = 0
+            for chunk in response.iter_content(chunk_size=1024*512):
+                if chunk:
+                    content.write(chunk)
+                    bytes_so_far += len(chunk)
+                    spinner.step(100 * float(bytes_so_far)/float(content_length))
+
+            # Extract the zip
+            release_temp_dir = self._get_temp_release_dir()
+            logger.debug("Extracting ArtifactTool to %s" % release_temp_dir)
+            f = zipfile.ZipFile(content)
+            f.extractall(path=release_temp_dir)
```

Comment:
consider a `try` around `f.extractall` through the os.rename, so you can delete the release_temp_dir folder if an exception occurs.

Reply:
Good idea, will do.